### ISSUE TITLE
Support CRA v4

### ISSUE THREAD

**Issue by JanJakes** [gsoft-inc/craco#205]
**CRA v4** was just released: https://github.com/facebook/create-react-app/blob/master/CHANGELOG.md

I gave it a quick try and I'm getting the following error at the moment:
```
craco:  *** Cannot find ESLint loader (eslint-loader). ***
```
(Didn't dig further into it though.)

**Comment by patricklafrance** [comment]
I guess a few things will break since a project like this is a major hack which break easily.

I will not personally invest time to support CRA v4 since I am not personally using it anymore. We do have a few projects at GSoft using CRACO so I still hope that someone from one of these teams will jump in to support CRA 4 but as of now, no one is willing to do it.

I will gladly review a PR if anyone is interested to submit one.

Thank you,

Patrick

**Comment by pbr1111** [comment]
The eslint-loader has been deprecated in favor of eslint-webpack-plugin, so it should be able to configure the plugin via the plugins property. The error is informative and even if we do not have any eslint configuration in the craco config it is displayed because by default the eslint property is filled with these values:
```
eslint: {
    mode: ESLINT_MODES.extends
}
```
Maybe it would be necessary to remove the default configuration of eslint or apply it only if the eslint-loader exists so as not to introduce a breaking change for those who use older versions of CRA.

**Comment by iwan-uschka** [comment]
I just read the whole thread so i am aware of

> I will not personally invest time to support CRA v4

but just want to put some extra bits in here. I just gave `react-scripts@4.0.0` a try and ran into the following issues:
- `cannot find module 'levenary'` => could be solved by manually installing `levenary`
- cannot resolve static files like fonts => https://github.com/facebook/create-react-app/issues/9870

I didn't notice any `ESLint` related error. My configuration:

- `react-scripts@4.0.0`
- `.eslintrc.js` containing
```
module.exports = {
  extends: 'react-app',
};
```
- no `ESLint` specific config in `craco.config.js`

Maybe maybe maybe this helps.

**Comment by patricklafrance** [comment]
Is eslint the only thing that breaks with `react-scripts@4.0.0` ?

**Comment by Thanaen** [comment]
@patricklafrance I personally get another error, with CRA v4:

`X:\xxx\node_modules\react-scripts\scripts\test.js:20
  throw err;
  ^

Error: craco: Cannot find Jest transform entry for Babel ^.+\.(js|jsx|ts|tsx)$ or ^.+\.(js|jsx)$. 
    at configureBabel (X:\xxx\node_modules\@craco\craco\lib\features\jest\merge-jest-config.js:35:27)
    at mergeJestConfig (X:\xxx\node_modules\@craco\craco\lib\features\jest\merge-jest-config.js:68:9)
    at proxy (X:\xxx\node_modules\@craco\craco\lib\features\jest\override.js:11:20)
    at Object.<anonymous> (X:\xxx\node_modules\react-scripts\scripts\test.js:76:5)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Module.require (internal/modules/cjs/loader.js:1025:19)
    at require (internal/modules/cjs/helpers.js:72:18)`

But I haven't made sure it's not an environment-related error, or a package-related error !

**Comment by pianomansam** [comment]
As there appear to be many issues with CRA v4, this bug report might be best broken out into individual reports to track the progress in fixing each.

I myself ran into the issue of changes to eslint due to the deprecation of eslint-loader. Specifically, my config disables eslint. I came up with a quick inline plugin to honor the `eslint.enable: false` setting.

```
  plugins: [
    {
      plugin: {
        overrideCracoConfig: ({ cracoConfig }) => {
          if (typeof cracoConfig.eslint.enable !== 'undefined') {
            cracoConfig.disableEslint = !cracoConfig.eslint.enable;
          }
          delete cracoConfig.eslint;
          return cracoConfig;
        },
        overrideWebpackConfig: ({ webpackConfig, cracoConfig }) => {
          if (
            typeof cracoConfig.disableEslint !== 'undefined' &&
            cracoConfig.disableEslint === true
          ) {
            webpackConfig.plugins = webpackConfig.plugins.filter(
              (instance) => instance.constructor.name !== 'ESLintWebpackPlugin',
            );
          }
          return webpackConfig;
        },
      },
    },
  ],
```

**Comment by justrealmilk** [comment]
I have the same issue as OP but I'm not as smart as the rest of y'all so I'm rolling back to 3.4.4 until...

**Comment by mormahr** [comment]
> @patricklafrance I personally get another error, with CRA v4:
> 
> `X:\xxx\node_modules\react-scripts\scripts\test.js:20
> throw err;
> ^
> 
> Error: craco: Cannot find Jest transform entry for Babel ^.+.(js|jsx|ts|tsx)$ or ^.+.(js|jsx)$.
> at configureBabel (X:\xxx\node_modules@craco\craco\lib\features\jest\merge-jest-config.js:35:27)
> at mergeJestConfig (X:\xxx\node_modules@craco\craco\lib\features\jest\merge-jest-config.js:68:9)
> at proxy (X:\xxx\node_modules@craco\craco\lib\features\jest\override.js:11:20)
> at Object. (X:\xxx\node_modules\react-scripts\scripts\test.js:76:5)
> at Module._compile (internal/modules/cjs/loader.js:1137:30)
> at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
> at Module.load (internal/modules/cjs/loader.js:985:32)
> at Function.Module._load (internal/modules/cjs/loader.js:878:14)
> at Module.require (internal/modules/cjs/loader.js:1025:19)
> at require (internal/modules/cjs/helpers.js:72:18)`
> 
> But I haven't made sure it's not an environment-related error, or a package-related error !

I submitted a PR to fix this. (#207)

**Comment by wojtow** [comment]
First of all, per the changelog on CRA v4, it may well be necessary to blow away your  node_modules and let yarn/npm rebuild it all.    For me, that fixed several of the  issues I was experiencing.

Secondly, make sure you're placing the blame appropriately.   Some of the issues you may be experiencing here are the result of the CRA v4 upgrade itself and not necessarily craco having issues with it.    If possible, try building with react-scripts instead of craco and see if the issues are still there.  I'm only using craco for some non-essential tweaks to the config, so I could test with react-scripts and found another issue was still there.  I dug through the issue reports with CRA and found a mitigation there. 

The only thing that remains for me now with craco is the eslint-loader message, which is just a warning that I'm ignoring for now.  My app seems to be building and working as expected otherwise at the moment.

**Comment by iwan-uschka** [comment]
I could solve [my issue](https://github.com/gsoft-inc/craco/issues/205#issuecomment-716514987)

> cannot resolve static files like fonts => facebook/create-react-app#9870

like described in https://github.com/facebook/create-react-app/issues/9870#issuecomment-722300637.

Can confirm `craco:  *** Cannot find ESLint loader (eslint-loader). ***` issue but it doesn't affect the app itself.

**Comment by devinrhode2** [comment]
@patricklafrance If you aren't using CRA anymore, what are you using?

**Comment by patricklafrance** [comment]
@devinrhode2 nothing, I have been developping full time a UI component library for the last year. The project I was working on are still using CRACO!

Sorry for the confusion.

**Comment by devuxer** [comment]
I am using TypeScript@4 and started having problems with react-scripts@3.4.4. I traced the problem to react-scripts depending on very old versions of @typescript-eslint/*, which were choking on some of the newer TypeScript features, so I tried upgrading to react-scripts@4.0, and now I find I'm not able to use CRACO (due to same error as OP). I need CRACO for Ant Design. So, all I can say is, I'm definitely hoping CRACO continues to be supported 🙏.

**Comment by kkulbae** [comment]
> **CRA v4** was just released: https://github.com/facebook/create-react-app/blob/master/CHANGELOG.md
> 
> I gave it a quick try and I'm getting the following error at the moment:
> 
> ```
> craco:  *** Cannot find ESLint loader (eslint-loader). ***
> ```
> 
> (Didn't dig further into it though.)

The eslint-loader has been deleted from create-react-app

https://github.com/facebook/create-react-app/commit/d07b7d025f5933710fcb01718617dbdf4bc54c33

Now eslint is being delivered through [eslint-webpack-plugin](https://github.com/webpack-contrib/eslint-webpack-plugin/tree/v2.1.0), not eslint-loader.

**Comment by iwan-uschka** [comment]
@kkulbae Thanks for the information, wasn't aware.. I researched how to solve this issue and i could find this plugin https://www.npmjs.com/package/craco-eslint-webpack-plugin but wasn't able to make it work. It somehow broke my whole craco configuration. What did you do?

**Comment by david29832** [comment]
it may be not the best syntax but i got it working with this:
```
const { BundleAnalyzerPlugin } = require('webpack-bundle-analyzer')
const CracoAntDesignPlugin = require('craco-antd')
const ESLintWebpackPlugin = require('eslint-webpack-plugin')
const WebpackBar = require('webpackbar')

module.exports = {
  eslint: false,
  webpack: {
    plugins: [
      new WebpackBar({ profile: true }),
      ...(process.env.NODE_ENV === 'development'
        ? [new BundleAnalyzerPlugin({ openAnalyzer: false })]
        : [])
    ]
  },
  plugins: [
    {
      plugin: CracoAntDesignPlugin,
      options: {
      }
    },
    {
      plugin: {
        overrideWebpackConfig: ({webpackConfig, pluginOptions }) => {
          webpackConfig.plugins.map((plugin)=> {
            if(plugin instanceof ESLintWebpackPlugin) {
              plugin.options.quiet = true
              plugin.options.emitWarning = true
            }
          })
          return webpackConfig
        }
      },
      options: {}
    }  ]
}

```

**Comment by JanJakes** [comment]
Everything works for me with Craco 5.8.0 and CRA 4.0.1 (including custom path aliases).
There is only the eslint warning (doing no harm).
Should I close this issue already?

**Comment by threepointone** [comment]
Yeah it's all working pretty well! But imho we should look to remove the eslint warning before closing this issue.

**Comment by threepointone** [comment]
I'll try to investigate this myself over the christmas break if no one else has a crack at it before then.

**Comment by Rocinante89** [comment]
Anyone using craco workbox plugin? Mine seems to have stopped generating a service worker after the update to 5.8.0 & CRA 4.0.1?

**Comment by JanJakes** [comment]
So I run into an issue – the eslint config seems to no longer work:
```js
eslint: {
  loaderOptions: {
    emitWarning: true,
  },
},
```

I achieved the same using an inline plugin as follows:
```js
{
  plugin: {
    overrideWebpackConfig: ({ webpackConfig }) => {
      const plugin = webpackConfig.plugins.find((plugin) => plugin.constructor.name === 'ESLintWebpackPlugin');
      plugin.options.emitWarning = true;
      return webpackConfig;
    }
  },
},
```

**Comment by patricklafrance** [comment]
Merged this PR https://github.com/gsoft-inc/craco/pull/219 from @alexasselin008 which fix CRACO ESLint feature for CR4 🥳 

It's released with https://github.com/gsoft-inc/craco/releases/tag/v6.0.0

While discussing this PR, we decided it was best if CRACO only support the latest MAJOR version of CRA to simplify the maintenance of CRACO and it's api. 

@alexasselin008 added a validation which run when CRACO start to ensure it's using a supported version of `react-script` otherwise an alternate version of CRACO is recommended depending on which `react-script` is installed. More info available [here](https://github.com/gsoft-inc/craco/blob/master/packages/craco/README.md#backward-compatibility).

With this PR, I believe it covers all the changes required to support CRA 4. It's been a great community achievement, thank you to everyone who participated in making this happen!
